rust/rsmgp-sys/src/value/mod.rs

```diff
@@ -0,0 +1,370 @@
+// Copyright (c) 2016-2021 Memgraph Ltd. [https://memgraph.com]
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ffi::{CStr, CString};
+
+use crate::edge::*;
+use crate::list::*;
+use crate::map::*;
+use crate::memgraph::*;
+use crate::mgp::*;
+use crate::path::*;
+use crate::result::*;
+use crate::vertex::*;
+// Required here, if not present tests linking fails.
+#[double]
+use crate::mgp::ffi;
+use mockall_double::double;
+
+/// # Safety
+/// TODO(gitbuda): Write section about safety.
+pub(crate) unsafe fn create_cstring(c_char_ptr: *const i8) -> MgpResult<CString> {
+    match CString::new(CStr::from_ptr(c_char_ptr).to_bytes()) {
+        Ok(v) => Ok(v),
+        Err(_) => Err(MgpError::UnableToMakeCString),
+    }
+}
+
+// NOTE on why mutable pointer to mgp_value has to be owned by this code.
+//
+// mgp_value used to return data is non-const, owned by the module code. Function to delete
+// mgp_value is non-const.
+// mgp_value containing data from Memgraph is const.
+//
+// `make` functions return non-const value that has to be deleted.
+// `get_property` functions return non-const copied value that has to be deleted.
+// `mgp_property` holds *const mgp_value.
+//
+// Possible solutions:
+//   * An enum containing *mut and *const can work but the implementation would also contain
+//   duplicated code.
+//   * A generic data type seems complex to implement https://stackoverflow.com/questions/40317860.
+//   * Holding a *const mgp_value + an ownership flag + convert to *mut when delete function has to
+//   be called.
+//   * Hold only *mut mgp_value... as soon as there is *const mgp_value, make a copy (own *mut
+//   mgp_value). The same applies for all other data types, e.g. mgp_edge, mgp_vertex.
+//   mgp_value_make_vertex accepts *mut mgp_vartex (required to return user data), but data from
+//   the graph is all *const T.
+//
+// The decision is to move on with a copy of mgp_value because it's cheap to make the copy.
+
+/// Useful to own `mgp_value` coming from / going into Memgraph as a result.
+///
+/// Underlying pointer object is going to be automatically deleted.
+#[derive(Debug)]
+pub struct MgpValue {
```

Comment:
If the memory would be stored inside MgpValue, then maybe the [From](https://doc.rust-lang.org/std/convert/trait.From.html) and/or [Into](https://doc.rust-lang.org/std/convert/trait.Into.html) traits could be utilized for providing idiomatic rust way for conversions between `MgpValue` and `Value`.

Reply:
Only `impl From<MgpValue> for Value` makes sense, but it's also not super nice to use it inside, e.g. vertex property, because `MgpValue` has to be created also there so it's I think more compact to just use `MgpValue`'s `to_value` method. + using this hides the `MgpResult` from getting retuned...

That's all for the library code, but for the user code, the `From` implementation makes sense :smile: